### DISCUSSION TITLE
locking before update

### DISCUSSION THREAD

**Discussion by stefanuytterhoeven** [custom-components/pyscript#340] [Q&A] [answered]
I'm using a persistent variable pyscript.counter.
But, several triggers can update that variable.
so, I'm asking myself, is there a way to lock that variable? So, it can't be update twice at the same time?
I've found a python example somewhere that uses :

from threading import Lock
lock = Lock()

lock.acquire()
counter +=1
lock.release()

I tried to implement that, but, evertime HA got unresponsive and had to turn my RPI off and on again. 
Maybe the cause was something else, but I guess this was not the way to go?

OR is there another way, a better way?

**Comment by craigbarratt** [comment, marked as answer]
Yes, there will be a race condition and it's great that you are aware of the issue.

You must use an async lock, not a thread lock.  Try this:
```python
import asyncio

counter_lock = asyncio.Lock()

def my_function():
    counter_lock.acquire()
    counter += 1
    counter_lock.release()
```
(note: untested)

**Reply by stefanuytterhoeven** [reply]
Seems to work. No race condition :)
When I have more time, I'll do some stress testing...

**Comment by stefanuytterhoeven** [comment]
Wow, thanks
Will try that!

Op wo 13 apr. 2022 20:10 schreef Craig Barratt ***@***.***>:
